### PR TITLE
FTP Operators new functions and bugfixes and deprecated COF

### PR DESCRIPTION
**Pull request by joergboe**
Here the changes as discussed in the tread of the last pull request.

### CODE REVIEW

**Comment by ddebrunner** [comment]
If you deprecating functionality the deprecation notice should inform what should be used instead.

File: com.ibm.streamsx.inet/com.ibm.streamsx.inet.ftp/FTPCommand/FTPCommand.xml [hunk 1]
```diff
@@ -53,14 +53,61 @@ The command string and the command arguments are received from port 0.
...
+            <description>This function returns a string with the url including schema.</description>
+            <prototype><![CDATA[rstring Url()]]></prototype>
+          </function>
+          <function>
+            <description>Deprecated: Get the number of executed ftp commands(in case of rename only one is counted)</description>
+            <prototype><![CDATA[uint32 NoTransfers()]]></prototype>
+          </function>
+          <function>
+            <description>Get the number of executed ftp commands(in case of rename only one is counted)</description>
+            <prototype><![CDATA[uint32 CommandCount()]]></prototype>
```
Comment (ddebrunner) [inline com.ibm.streamsx.inet/com.ibm.streamsx.inet.ftp/FTPCommand/FTPCommand.xml]: This looks wrong, CommandCount already exists at line 61

File: com.ibm.streamsx.inet/com.ibm.streamsx.inet.ftp/FTPCommand/FTPCommand.xml [hunk 2]
```diff
@@ -199,6 +246,26 @@ This optionally parameter controls when the closing of connection after transfer
...
+        <cardinality>1</cardinality>
+        <portScope><port>0</port></portScope>
+      </parameter>
+      <parameter>
+        <name>transferTimeout</name>
+        <description>This is the maximum time in seconds that you allow the libcurl transfer operation to take. Normally, name lookups can take a considerable time. Default timeout is 0 (zero) which means it never times out.</description>
+        <optional>true</optional>
+        <rewriteAllowed>true</rewriteAllowed>
+        <expressionMode>Expression</expressionMode>
+        <type>int32</type>
```
Comment (ddebrunner) [inline com.ibm.streamsx.inet/com.ibm.streamsx.inet.ftp/FTPCommand/FTPCommand.xml]: All other timeouts in operators are float64, can this be consistent with the common practice.

File: com.ibm.streamsx.inet/com.ibm.streamsx.inet.ftp/FTPCommand/FTPCommand.xml [hunk 3]
```diff
@@ -199,6 +246,26 @@ This optionally parameter controls when the closing of connection after transfer
         <cardinality>1</cardinality>
       </parameter>
       <parameter>
+        <name>connectionTimeout</name>
+        <description>This is the maximum time in seconds that you allow the connection to the server to take. This only limits the connection phase, once it has connected, this option is of no more use. Set to zero to switch to the default built-in connection timeout - 120 seconds. See also the `transferTimeout` parameter.</description>
+        <optional>true</optional>
+        <rewriteAllowed>true</rewriteAllowed>
+        <expressionMode>Expression</expressionMode>
+        <type>int32</type>
```
Comment (ddebrunner) [inline com.ibm.streamsx.inet/com.ibm.streamsx.inet.ftp/FTPCommand/FTPCommand.xml]: All other timeouts in operators are float64, can this be consistent with the common practice.

File: com.ibm.streamsx.inet/com.ibm.streamsx.inet.ftp/FTPCommand/FTPCommand_cpp.cgt [hunk 4]
```diff
@@ -96,7 +134,7 @@ void MY_OPERATOR::prepareToShutdown() {
 
 // Tuple processing for non mutating ports
 void MY_OPERATOR::process(Tuple const & tuple, uint32_t port) {
-	SPLLOG(L_TRACE, "process(" << tuple << ", " << port << ")", debugAspect);
+	SPLAPPTRC(L_TRACE, "process(" << tuple << ", " << port << ")", debugAspect);
```
Comment (ddebrunner) [inline com.ibm.streamsx.inet/com.ibm.streamsx.inet.ftp/FTPCommand/FTPCommand_cpp.cgt]: Should this be protected with a test to see if we are logging at this level.

Also is this actually needed, does Streams trace tuples at fine levels anyway?

File: com.ibm.streamsx.inet/com.ibm.streamsx.inet.ftp/FTPPutFile/FTPPutFile.xml [hunk 5]
```diff
@@ -229,6 +257,27 @@ This optionally parameter controls when the closing of connection after transfer
...
+        <portScope><port>0</port></portScope>
+      </parameter>
+      <parameter>
+        <name>transferTimeout</name>
+        <description>This is the maximum time in seconds that you allow the libcurl transfer operation to take. Normally, name lookups can take a considerable time
+and limiting operations to less than a few minutes risk aborting perfectly normal operations. Default timeout is 0 (zero) which means it never times out.</description>
+        <optional>true</optional>
+        <rewriteAllowed>true</rewriteAllowed>
+        <expressionMode>Expression</expressionMode>
+        <type>int32</type>
```
Comment (ddebrunner) [inline com.ibm.streamsx.inet/com.ibm.streamsx.inet.ftp/FTPPutFile/FTPPutFile.xml]: float64

File: com.ibm.streamsx.inet/com.ibm.streamsx.inet.ftp/FTPPutFile/FTPPutFile.xml [hunk 6]
```diff
@@ -229,6 +257,27 @@ This optionally parameter controls when the closing of connection after transfer
         <cardinality>1</cardinality>
       </parameter>
       <parameter>
+        <name>connectionTimeout</name>
+        <description>This is the maximum time in seconds that you allow the connection to the server to take. This only limits the connection phase, once it has connected, this option is of no more use. Set to zero to switch to the default built-in connection timeout - 120 seconds. See also the `transferTimeout` parameter.</description>
+        <optional>true</optional>
+        <rewriteAllowed>true</rewriteAllowed>
+        <expressionMode>Expression</expressionMode>
+        <type>int32</type>
```
Comment (ddebrunner) [inline com.ibm.streamsx.inet/com.ibm.streamsx.inet.ftp/FTPPutFile/FTPPutFile.xml]: float64

File: com.ibm.streamsx.inet/com.ibm.streamsx.inet.ftp/FTPReader/FTPReader.xml [hunk 7]
```diff
@@ -274,6 +334,27 @@ This optionally parameter controls when the closing of connection after transfer
...
+        <portScope><port>0</port></portScope>
+      </parameter>
+      <parameter>
+        <name>transferTimeout</name>
+        <description>This is the maximum time in seconds that you allow the libcurl transfer operation to take. Normally, name lookups can take a considerable time
+and limiting operations to less than a few minutes risk aborting perfectly normal operations. Default timeout is 0 (zero) which means it never times out.</description>
+        <optional>true</optional>
+        <rewriteAllowed>true</rewriteAllowed>
+        <expressionMode>Expression</expressionMode>
+        <type>int32</type>
```
Comment (ddebrunner) [inline com.ibm.streamsx.inet/com.ibm.streamsx.inet.ftp/FTPReader/FTPReader.xml]: float64

File: samples/PutFile/com.ibm.streamsx.inet.ftp.sample/PutFile.spl [hunk 8]
```diff
@@ -4,7 +4,7 @@
 namespace com.ibm.streamsx.inet.ftp.sample;
 
 use com.ibm.streamsx.inet.ftp::*;
-//use com.ibm.streams.teda.file::*;
+//use com.ibm.streams.teda.file::rename;
```
Comment (ddebrunner) [inline samples/PutFile/com.ibm.streamsx.inet.ftp.sample/PutFile.spl]: Can we remove this commented-out use statemetns

File: com.ibm.streamsx.inet/com.ibm.streamsx.inet.ftp/FTPCommand/FTPCommand.xml (hunk 1 above)
Comment (joergboe) [inline com.ibm.streamsx.inet/com.ibm.streamsx.inet.ftp/FTPCommand/FTPCommand.xml]: This is not wrong. In line 61 the function declaration is group 'GetData'. In line 100 the declaration is function group 'GetError'.

File: com.ibm.streamsx.inet/com.ibm.streamsx.inet.ftp/FTPCommand/FTPCommand.xml (hunk 2 above)
Comment (joergboe) [inline com.ibm.streamsx.inet/com.ibm.streamsx.inet.ftp/FTPCommand/FTPCommand.xml]: The underlying curl library defines the interface for timeouts in seconds granularity only.
So it makes no sense to define float values for time-outs and add a comment that the values are rounded (or cut) to seconds operator level. The definition of a int implies that no fractions are possible/usable.

File: com.ibm.streamsx.inet/com.ibm.streamsx.inet.ftp/FTPCommand/FTPCommand.xml (hunk 3 above)
Comment (joergboe) [inline com.ibm.streamsx.inet/com.ibm.streamsx.inet.ftp/FTPCommand/FTPCommand.xml]: dito

File: com.ibm.streamsx.inet/com.ibm.streamsx.inet.ftp/FTPPutFile/FTPPutFile.xml (hunk 5 above)
Comment (joergboe) [inline com.ibm.streamsx.inet/com.ibm.streamsx.inet.ftp/FTPPutFile/FTPPutFile.xml]: see above

File: com.ibm.streamsx.inet/com.ibm.streamsx.inet.ftp/FTPPutFile/FTPPutFile.xml (hunk 6 above)
Comment (joergboe) [inline com.ibm.streamsx.inet/com.ibm.streamsx.inet.ftp/FTPPutFile/FTPPutFile.xml]: see above

File: com.ibm.streamsx.inet/com.ibm.streamsx.inet.ftp/FTPReader/FTPReader.xml (hunk 7 above)
Comment (joergboe) [inline com.ibm.streamsx.inet/com.ibm.streamsx.inet.ftp/FTPReader/FTPReader.xml]: see above

File: com.ibm.streamsx.inet/com.ibm.streamsx.inet.ftp/FTPCommand/FTPCommand_cpp.cgt (hunk 4 above)
Comment (joergboe) [inline com.ibm.streamsx.inet/com.ibm.streamsx.inet.ftp/FTPCommand/FTPCommand_cpp.cgt]: 'Should this be protected with a test to see if we are logging at this level.'
No the the C macro SPLAPPTRC achieves that for us. SPLAPPTRC call makro APPTRC.

'does Streams trace tuples at fine levels anyway?' No at least not in the trace file. Here is a sample with trace level:
23 May 2016 10:33:12.918 [9326] TRACE #splapptrc,J[1],P[11],ResultStream,FTPCommand M[ResultStream.cpp:process:66]  - process({command="mkdir",file="archive",file2=""}, 0)
23 May 2016 10:33:12.918 [9326] TRACE #splapptrc,J[1],P[11],ResultStream,FTPWrapper,FTPCommand M[FTPWrapper.cpp:perform:366]  - Perform FTP action url="ftp://localhost/ftptry/"
23 May 2016 10:33:12.918 [9326] TRACE #splapptrc,J[1],P[11],ResultStream,FTPWrapper,FTPCommand M[FTPWrapper.cpp:headerCallback:875]  - header has=1 count=40 bytes
23 May 2016 10:33:12.918 [9326] DEBUG #splapptrc,J[1],P[11],ResultStream,FTPWrapper,FTPCommand M[FTPWrapper.cpp:headerCallback:880]  - ftp command returns:"550 Create directory operation failed.\r\n"
23 May 2016 10:33:12.919 [9326] ERROR #splapptrc,J[1],P[11],ResultStream,FTPCommand M[ResultStream.cpp:sendError:151]  - "libcurl perform operation failed! curl says: Quote command returned error url:ftp://localhost/ftptry/ command:mkdir archive "
23 May 2016 10:33:12.919 [9326] TRACE #splapptrc,J[1],P[11],ResultStream,FTPCommand M[ResultStream.cpp:process:66]  - process({command="rename",file="file.txt",file2="archive/file.txt

**Pull request merged**